Add Subscriptions Migrator

leonardola opened the pull request:
This PR adds a way to migrate Subscriptions


### Test instructions
- Checkout to this PR's branch
- Create two orders with the same product
- Update the orders creation date via the /wp-admin order edit page to `2023-01-07` and `2023-01-08`
- Add a `_order_number` meta and set it to `11664` in the oldest order
- Add a `_order_number` meta and set it to `11665` in the newest order
- Download these files
[orders.json](https://github.com/woocommerce/migrator-cli/files/13044179/orders.json)
[subscriptions.json](https://github.com/woocommerce/migrator-cli/files/13044180/subscriptions.json)
- Run `wp migrator skio_subscriptions --subscriptions_export_file=<file_path>  --orders_export_file=<file_path>`
- Make sure the subscription was created and uses the shipping and billing data from the newest order
- Change the shipping and billing address in the newest order
- Run `wp migrator skio_subscriptions --subscriptions_export_file=<file_path>  --orders_export_file=<file_path>` again
- Make sure the subscription got updated
- Change the status from `ACTIVE` to `CANCELLED` in the subscription.json file
- Run `wp migrator skio_subscriptions --subscriptions_export_file=<file_path>  --orders_export_file=<file_path>` again
- Make sure the subscription is now cancelled
- Run any other tests you think need to run

Commit:
Started to create the migrator for subscriptions

Commit:
Fixed payment method for subscription

nagpai reviewed (commented):
Early test - Tested manually with an actual order imported from the live site. It works as expected

Commit:
now it's possible to import subscriptions from skio

Commit:
Moved the subscription import function to it's own file

Commit:
Removed unecessary skio token from config file

Commit:
Added Readme and docblock

Commit:
cs fixes

leonardola commented:
@nagpai I've updated the code to use the export files instead of the api as not all data needed is available there.

I've updated the PR description. 

Could you please test again?

nagpai commented on includes/utils.php:
This is a nice refactor, moving the util functions to a separate class!

nagpai commented:
I did a manual test and was unable to see the subscriptions getting created. Here are the steps I followed: 

1. I used a LocalWP instance for testing
2. I created the orders as per the instructions, and `_order_number` meta using Adminer tool, by adding the relevant entries to the `wp_postmeta` table. I hope this is the correct way to do it. Let me know if there is any other step needed. 
3. I moved the JSON files to ./tmp path and executed this command 
`wp migrator skio_subscriptions --subscriptions_export_file=./tmp/subscriptions.json --orders_export_file=./tmp/orders.json`

also tested this command with the path in single quotes. Did not work either. 

4. The command executed without showing any errors. It showed. 

```
wp migrator skio_subscriptions --subscriptions_export_file=./tmp/subscriptions.json --orders_export_file=./tmp/orders.json       
Adding Subscription ids to orders
Creating Subscriptions
Done
```
5. When I checked the subscriptions section of the site, I could not find any. 

I made sure I had All products for subscriptions extension and Woo Subscriptions active.

nagpai commented:
I ran the steps with absolute paths instead of relative for the JSON files, and it worked correctly! ( Thanks for the tip @leonardola ) 

✅  Subscription created correctly with orders correctly connected to it. First order is set as Parent, and second order is set as renewal order.
🟨  - Change of address reflects correctly on the subscription. However running the command again seems to add an additional product at the end of it. Meaning subsequent renewals will have twice the number of products. See screenshot: 

![image](https://github.com/woocommerce/migrator-cli/assets/4162931/ef0cca54-a354-443d-bffe-3ac639c26039)

🟨 - Changing status of the sub to "CANCELLED" updates it correctly. However running the command adds an additional product to it. 

In my earlier try, the command did not explicitly indicate no JSON / data imported. It will be nice to show an error/ warning message mentioning no data. Suggested a [snippet below](https://github.com/woocommerce/migrator-cli/pull/3#pullrequestreview-1701736752).

Commit:
added checks to warn about missing files

nagpai commented on includes/subscriptions.php:
```diff
@@ -0,0 +1,252 @@
+		$subscription->add_item( $new_item );
+	}
+
+	private function get_data_from_file( $file )
```
```suggestion
	private function get_data_from_file( $file )
	{
		if ( ! file_exists( $file ) ) {
			throw new \Exception( 'File not found: ' . $file );
		}
		return json_decode(file_get_contents( $file ), true);
	}
```

Result: 

<img width="700" alt="image" src="https://github.com/woocommerce/migrator-cli/assets/4162931/9f7eebbb-639f-40e7-8578-e4107c78f4ef">

Commit:
fixed bug that would duplicate order items when updating a subscription

leonardola commented:
@nagpai I've fixed the issues you found. Could you please take another look?

Missing files: ededc7ec42ab6a126b81658035d0d2de410cc216
Duplicate items: ce8d9a23e1ae72c10623b8c2cf6f50888709e377

nagpai commented on includes/subscriptions.php:
```diff
@@ -0,0 +1,262 @@
+
+	private function get_data_from_file( $file )
+	{
+		if ( ! is_file( $file ) ) {
```
I see the `File not found` error, even if I provide the correct absolute paths. This may be because we are using  `is_file`, that does not work with absolute paths. 

We may need to use something like this: 

```suggestion
		private function get_data_from_file( $file )
	{
		$file_contents = file_get_contents( $file );

		if ( $file_contents === false ) {
			WP_CLI::line( WP_CLI::colorize( '%RFile not found: ' . $file ) );
			die();
		}

		return json_decode( $file_contents, true);
	}
	
	
```

nagpai replied:
If i change this and test, the duplication of items still exists when I re-run the command.

leonardola replied:
Duplication problem fixed with 1ce0178188ef44a5dbe92508cc3e79a4ca3760e8

leonardola replied:
I tried running the command as follow:

`wp migrator skio_subscriptions --subscriptions_export_file=/Users/leonardoalbuquerque/Local/taika/app/public/wp-content/plugins/migrator-cli/.tmp/subscriptions.json  --orders_export_file=/Users/leonardoalbuquerque/Local/taika/app/public/wp-content/plugins/migrator-cli/.tmp/orders.json`

And it found the file. 

What did not work was to run it like this:

`wp migrator skio_subscriptions --subscriptions_export_file=../.tmp/subscriptions.json  --orders_export_file=../.tmp/orders.json`

nagpai replied:
Ah it did not work for me since I used URL of the localhost. and is_file does not work with URL paths. I will try again with the local path. We may need to add this in the Readme.

nagpai replied:
@leonardola  The duplication issue seems to be resolved. The command also works correctly with the full local path.    However, for these steps I do not see the order addresses changing for these testing steps shared by you:

```
* Change the shipping and billing address in the newest order
* Run wp migrator skio_subscriptions --subscriptions_export_file=<file_path>  --orders_export_file=<file_path> again
* Make sure the subscription got updated
```

If it works fine at your end, it could be very likely my local setup.

leonardola replied:
> However, for these steps I do not see the order addresses changing for these testing steps shared by you:

Sometimes the wrong order was being used to create/update the subscription. Fixed with: 872b8f1d3e64336312061df228a55415bfa4dd50

nagpai reviewed (changes requested):
Left some comments. Thanks.

Commit:
Fixed bug that would still let duplicated items to be added to subscriptions

Commit:
Now the oldest order is set as the parent order in subscriptions

Commit:
added output with the Woo Subscription id
